cpp/src/arrow/array/data.h

```diff
@@ -434,6 +434,21 @@ struct ARROW_EXPORT ArraySpan {
     return GetValues<T>(i, this->offset);
   }
 
+  // Access a buffer's data as a span
+  template <typename T>
+  util::span<const T> GetSpan(int i) const {
+    return util::span(buffers[i].data_as<T>(),
+                      static_cast<size_t>(buffers[i].size) / sizeof(T))
+        .subspan(offset);
+  }
+
+  template <typename T>
+  util::span<T> GetSpan(int i) {
+    return util::span(buffers[i].mutable_data_as<T>(),
+                      static_cast<size_t>(buffers[i].size) / sizeof(T))
+        .subspan(offset);
```

Comment:
Should this length be the array length instead? Considering that it's possible (I believe) to have buffers that are longer than `(offset+length)*sizeof(T)`?

@bkietz @pitrou what are the invariants followed by the buffer lengths for valid arrays? Are they always perfectly sliced to fit `offset+length` values or can they be longer?

Reply:
The buffers may be longer.

It'd be nicer if the spans could be sliced to reflect the array length, but what about offset buffers? It seems we'd need to incorporate layout info. Doable, but maybe not worth it here?

Reply:
Isn't that just `.subspan(this->offset, this->length)` for the functions defined in `ArraySpan` like the one under this comment?

Reply:
Alright, by "offset buffer" you mean the buffers of `List` arrays, not buffers in arrays with offsets. We would need a switch on type and buffer id which has the potential to be slow and a pain to maintain.

What if we then ask for the length as a parameter to `GetSpan`? Then all we have to do is assert that it's `<=` the buffer size divided by the `sizeof(T)` (plus the array offset handling).

Reply:
I think adding an explicit parameter for length would be fine but maybe wouldn't add much value:
```diff
-data.GetSpan<int32_t>(1).subspan(0, 50);
+data.GetSpan<int32_t>(1, /*length=*/50);
```

Maybe instead we could just add a `bool one_extra` parameter? 
```diff
-data.GetSpan<int32_t>(1).subspan(0, 50);
+data.GetSpan<int32_t>(1, /*one_extra=*/true); // offsets buffer
```

Also, it's worth noting that these will (I think) only rarely be used on a buffer other than 1; it'll never be correct to use these on a null bitmap and I think list view and dense union are the only layouts which have buffers beyond 1 which aren't raw bytes

Reply:
I think the value-add comes from a reduced risk of code that doesn't account for the fact that the returned `span` has (sometimes, very rarely) length greater than the array length and even-worse: the fact that the data after the end of the length could be non-sensical (e.g. zeroes at the end of an offsets buffer).

We risk having compute kernel code that processes these spans without issue until one that's greater than the array length comes up. Taking the length as parameter, drastically removes this branch in the program graph: the span always has the array length instead of possibly also having a greater length (in rare cases!).

Regarding this:

```diff
-data.GetSpan<int32_t>(1).subspan(0, 50);
+data.GetSpan<int32_t>(1, /*length=*/50);
```

For what I'm proposing these are not equivalent. I want to assert when the passed length is greater than the available buffer size instead of truncating like `subspan` does. I would also consider debug-build asserts that take into account the array type and buffer index -- something that `span::subspan()` can't do.

```
  constexpr span subspan(size_t offset) const {
    if (offset > size_) return {data_, data_};
    return {data_ + offset, size_ - offset};
  }
```

Regarding this suggestion:

```
-data.GetSpan<int32_t>(1).subspan(0, 50);
+data.GetSpan<int32_t>(1, /*one_extra=*/true); // offsets buffer
```

I think this would be a very ergonomic API that could in turn be defined in terms of the `GetSpan` implementation that takes a `length`.

> it'll never be correct to use these on a null bitmap

Definitely. We should assert (or at least document) that `i > 0`.

A final observation: we already have functions that return the base pointer of the buffer (shifted by offset), if we are going to provide span-returning functions, we better make the length of the span maximally useful since that's exactly what spans add compared to raw pointers -- a length.

Reply:
@jsjtxietian TL;DR: please ensure the GetSpan functions

- accept an explicit length parameter or at least a one_extra parameter
- assert that the output length is in fact available within the indicated buffer
- add tests checking that the output length is correct